framework/src/materials/Material.C

```diff
@@ -171,8 +175,26 @@ Material::getOutputs()
 void
 Material::computeProperties()
 {
-  for (_qp = 0; _qp < _qrule->n_points(); ++_qp)
+  // If this Material has the _constant_on_elem flag set, we take the
+  // value computed for _qp==0 and use it at all the quadrature points
+  // in the Elem.
+  if (_constant_on_elem)
+  {
+    // Compute MaterialProperty values at the first qp.
+    _qp = 0;
     computeQpProperties();
+
+    // Copy the first value computed over to the others.
+    MaterialProperties & props = _material_data->props();
```

Comment:
`_supplied_props` is the right place. I'd say loop over that to get the property names, transform those into `prop_ids` and then use that to index into `props()`.  The transformation method sits in `MaterialPropertyStorage`, so you have to go get that from `FEProblemBase`.